### PR TITLE
New ruler alerts that ignore user-errors

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR adds new metrics for tracking total number of queries and push requests to ingesters from rulers, and failed queries and push requests. Unlike existing `cortex_prometheus_rule_evaluation_failures_total`, failures are only counted for "real" failures and not user failures (hitting limits, duplicate samples, etc.). The idea is to use these new metrics in alerts, to avoid being alerted on issues that users can already solve themselves.

This PR also moves `errorTranslateQueryable` from `api` package to `querier` package, for simpler reuse.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ruler/compat.go [hunk 1]
```diff
@@ -60,9 +66,25 @@ func (a *pusherAppender) AppendExemplar(_ uint64, _ labels.Labels, _ exemplar.Ex
...
+		reportError := true
+
+		if resp, ok := httpgrpc.HTTPResponseFromError(err); ok && resp.Code/100 == 4 {
+			// Don't report errors that ended with 4xx HTTP status code (series limits, duplicate samples, out of order, etc.)
+			reportError = false
+		}
+
+		if reportError {
+			a.failedWrites.Inc()
+		}
```
Comment (jtlisi) [inline pkg/ruler/compat.go:85]: **suggestion:** If I'm not mistaken, this should be equivalent according to de morgans law and a bit tidier: 
```suggestion
		// Don't report errors that ended with 4xx HTTP status code (series limits, duplicate samples, out of order, etc.)
		if resp, ok := httpgrpc.HTTPResponseFromError(err); !ok || resp.Code/100 != 4 {
			a.failedWrites.Inc()
		}
```

@pstibrany WDYT?

File: pkg/ruler/compat.go [hunk 2]
```diff
@@ -129,12 +187,30 @@ type RulesManager interface {
...
+
+	totalQueries := promauto.With(globalReg).NewCounter(prometheus.CounterOpts{
+		Name: "cortex_ruler_queries_total",
+		Help: "Number of queries executed by ruler.",
+	})
+	failedQueries := promauto.With(globalReg).NewCounter(prometheus.CounterOpts{
+		Name: "cortex_ruler_queries_failed_total",
+		Help: "Number of failed queries by ruler.",
+	})
+
```
Comment (jtlisi) [inline pkg/ruler/compat.go:208]: **question:** why use  the name globalReg here? Is this to let let the caller know not to use a per-tenant reg?

```suggestion
func DefaultTenantManagerFactory(cfg Config, p Pusher, q storage.Queryable, engine *promql.Engine, overrides RulesLimits, reg prometheus.Registerer) ManagerFactory {
	totalWrites := promauto.With(reg).NewCounter(prometheus.CounterOpts{
		Name: "cortex_ruler_write_requests_total",
		Help: "Number of write requests to ingesters.",
	})
	failedWrites := promauto.With(reg).NewCounter(prometheus.CounterOpts{
		Name: "cortex_ruler_write_requests_failed_total",
		Help: "Number of failed write requests to ingesters.",
	})

	totalQueries := promauto.With(reg).NewCounter(prometheus.CounterOpts{
		Name: "cortex_ruler_queries_total",
		Help: "Number of queries executed by ruler.",
	})
	failedQueries := promauto.With(reg).NewCounter(prometheus.CounterOpts{
		Name: "cortex_ruler_queries_failed_total",
		Help: "Number of failed queries by ruler.",
	})

```

File: pkg/ruler/compat_test.go [hunk 3]
```diff
@@ -110,3 +115,129 @@ func TestPusherAppendable(t *testing.T) {
 		})
 	}
 }
+
+func TestPusherErrors(t *testing.T) {
```
Comment (jtlisi) [inline pkg/ruler/compat_test.go:119]: Nice tests

**Review by jtlisi — APPROVED** [review]
LGTM

File: pkg/ruler/compat.go (hunk 1 above)
Comment (pstibrany) [reply pkg/ruler/compat.go:85]: Looks good, you know your boolean algebra well 😄

File: pkg/ruler/compat.go (hunk 2 above)
Comment (pstibrany) [reply pkg/ruler/compat.go:208]: > **question:** why use the name globalReg here? Is this to let let the caller know not to use a per-tenant reg?

Yes. I wanted to make it explicit that registry here is not the same as registry used by function in the return statement (which is per-tenant). But I guess it's clear enough already.

File: pkg/ruler/compat_test.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ruler/compat_test.go:119]: Thank you.

**Review by pracucci — APPROVED** [review]
Good job! 👏

**Pull request merged**